### PR TITLE
Improve `hashes::Error`

### PR DESCRIPTION
**Pull request by tcharding**
We are trying to make error types stable on the way to v1.0

The current `hashes::Error` is a "general" enum error type with a single variant, better to use a struct and make the error usecase specific.

Improve the `hashes::Error` by doing:

- Make it a struct
- Rename to `FromSliceError`
- Move it to the crate root (remove `error` module)

Includes usage in `bitcoin`.

### CODE REVIEW

**Comment by Kixunil** [comment]
Since Rust now supports converting slices to arrays it's tempting to remove `from_slice` methods together with this error completely. However there are still some advantages of having these:

* Shorter to write
* Contains more context compared to `core::array::TryFromSliceError`

So I'm not really sure what to do.

**Comment by apoelstra** [comment]
I think we should keep move forward with this. It's a low-cost thing to maintain (slice conversions will basically never change, nor do they interact badly with other features) so worst-case it will wind up as a sorta vestigial part of our API.

File: bitcoin/src/psbt/error.rs [hunk 1]
```diff
@@ -57,7 +57,7 @@ pub enum Error {
     /// Unable to parse as a standard sighash type.
     NonStandardSighashType(u32),
     /// Parsing errors from bitcoin_hashes
-    HashParse(hashes::Error),
+    HashParse(hashes::FromSliceError),
```
Comment (Kixunil) [inline bitcoin/src/psbt/error.rs:60]: Calling this one `HashParse` looks kinda strange now but this is for a different PR.

File: hashes/src/lib.rs [hunk 2]
```diff
@@ -219,6 +217,24 @@ pub trait Hash:
     fn all_zeros() -> Self;
 }
 
+/// Attempted to create a hash from an invalid length slice.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct FromSliceError {
+    expected: usize,
+    got: usize,
```
Comment (Kixunil) [inline hashes/src/lib.rs:224]: Might be useful to make these `pub` and make the struct `#[non_exhaustive]`.

**Review by Kixunil — APPROVED** [review]
ACK 06afd52a12b74a600747d1563786fbf53cd37018

This is a great improvement!

File: bitcoin/src/psbt/error.rs (hunk 1 above)
Comment (tcharding) [reply bitcoin/src/psbt/error.rs:60]: Would be nice to come up with a naming convention for error variants that wrap single usecase errors, I can't think of an exact example right now but over the last few weeks I've been unhappy with my naming. Will mention again later.

File: hashes/src/lib.rs (hunk 2 above)
Comment (tcharding) [reply hashes/src/lib.rs:224]: Made mental note to play with this idea and check usability of such an error type. Thanks

File: bitcoin/src/psbt/error.rs (hunk 1 above)
Comment (Kixunil) [reply bitcoin/src/psbt/error.rs:60]: I'd just call it `Hash` or `InvalidHash`, since this is not parsing.

Comment (tcharding) [reply bitcoin/src/psbt/error.rs:60]: #1888

**Pull request merged**